annnufan/assignment2/node.cpp

```diff
@@ -0,0 +1,96 @@
+#include <iostream>
+#include <algorithm>
+
+
+class node {
+	int key;
+public:
+	node(int x, node* l = nullptr, node* r = nullptr) : key(x), left_node(l), right_node(r){}
+	~node() {
+		delete left_node;
+		delete right_node;
+	}
+
+	void set_value(int x) {
+		key = x;
+	}
+
+	int get_value() {
+		return key;
+	}
+
+	node* left_node, *right_node, *parent_node;
+};
+
+bool find_key_and_print_way_to_root(int x, node* v) {
+	if (v == nullptr) {
+		return false;
+	}
+	if (x == v->get_value()) {
+		return true;
+	}
+	if (find_key_and_print_way_to_root(x, v->left_node) || find_key_and_print_way_to_root(x, v->right_node)) {
+		std::cout << v->get_value() << ' ';
+		return true;
+	}
+	return false;
+}
+
+void print_way_to_root_for_key(int x, node* v) {
+	if (find_key_and_print_way_to_root(x, v)) {
+		std::cout << "\n";
+	} else {
+		std::cout << "Couldn't find key\n";
+	}
+}
+
+node* find_node(int x, node* v, node* p) {
+	if (v == nullptr) {
+		return nullptr;
+	}
+	v->parent_node = p;
+	if (x == v->get_value()) {
+		return v;
+	}
+	node* u = find_node(x, v->left_node, v);
+	if (u != nullptr) {
+		return u;
+	}
+	u = find_node(x, v->right_node, v);
+	return u;
+}
+
+int count_high(node* v) {

```

Comment:
This function assumes that `parent_node` is set (which is not always the case at this moment).
So you can add a comment that clarifies this:
```C++
// Assumes that parent_node is set for this node and its parents.
int count_high(...
```

But better: see tip above to make sure that `parent_node` is always set, so this function always works as expected.